examples/ilu-preconditioned-solver/ilu-preconditioned-solver.cpp

```diff
@@ -0,0 +1,118 @@
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+
+int main(int argc, char *argv[])
+{
+    // Some shortcuts
+    using vec = gko::matrix::Dense<>;
+    using mtx = gko::matrix::Csr<>;
+    using gmres = gko::solver::Gmres<>;
+
+    // Print version information
+    std::cout << gko::version_info::get() << std::endl;
+
+    // Figure out where to run the code
+    std::shared_ptr<gko::Executor> exec;
+    if (argc == 1 || std::string(argv[1]) == "reference") {
+        exec = gko::ReferenceExecutor::create();
+    } else if (argc == 2 && std::string(argv[1]) == "omp") {
+        exec = gko::OmpExecutor::create();
+    } else if (argc == 2 && std::string(argv[1]) == "cuda" &&
+               gko::CudaExecutor::get_num_devices() > 0) {
+        exec = gko::CudaExecutor::create(0, gko::OmpExecutor::create());
+    } else {
+        std::cerr << "Usage: " << argv[0] << " [executor]" << std::endl;
+        std::exit(-1);
+    }
+
+    // Read data
+    auto A = share(gko::read<mtx>(std::ifstream("data/A.mtx"), exec));
+    auto b = gko::read<vec>(std::ifstream("data/b.mtx"), exec);
+    auto x = gko::read<vec>(std::ifstream("data/x0.mtx"), exec);
+
+    // Generate incomplete factors using ParILU
+    auto par_ilu_fact = gko::factorization::ParIlu<>::build().on(exec);
+    // Generate concrete factorization for input matrix
+    auto par_ilu = par_ilu_fact->generate(A);
+
+    // Generate an ILU preconditioner factory by setting lower and upper
+    // triangular solver - in this case the exact triangular solves
+    auto ilu_pre_factory =
+        gko::preconditioner::Ilu<gko::solver::LowerTrs<>,
+                                 gko::solver::UpperTrs<>, false>::build()
+            .on(exec);
+
+    // Use incomplete factors to generate ILU preconditioner
+    auto ilu_preconditioner = ilu_pre_factory->generate(gko::share(par_ilu));
+
+    // Use preconditioner inside GMRES solver factory
+    // Generating a factory makes sense if there are several systems for which
```

Comment:
Why does generating a "factory" make sense when you could reuse an ILU preconditioner?
Did you mean "using a generated preconditioner in the factory makes sense if several systems are expected to be effective with it" or similar?

Reply:
I tried to answer the question in the updated comment: a prec+solver combination might be effective for several problems, this is why we may generate a factory of this kind.